### PR TITLE
BUG: spatial.distance.jensenshannon: fix returning NaN for some similar distributions

### LINKED ISSUE #20083: BUG: Jensen shannon returns NaN (takes sqrt of small negative number)
### Describe your issue.

For some cherry-picked pairs of distributions that are near one another, jensenshannon returns "nan".

The cause is that the code takes the sqrt of a small negative number [here](https://github.com/scipy/scipy/blob/9d08eb58205fb669ff72c60080b18be61b14d8e1/scipy/spatial/distance.py#L1272).

You might fix it with

`js = np.clip(js, a_min=0.0, a_max=None)`

before the sqrt

Thanks for scipy!

### Reproducing Code Example

```python
import numpy as np
from scipy.spatial.distance import jensenshannon
import struct

c1 = np.array([0.027501475, 0.055202297], dtype='float32')
c2 = np.array([0.027537677, 0.055334348], dtype='float32')
print("c1:", c1)
print("c2:", c2)
print("type:", c1.dtype, "shape:", c1.shape)

# returns nan!
print("first jsd:", jensenshannon(c1, c2))

# does not return nan
c1[0]=c1[0]+0.000000001
print("second jsd:", jensenshannon(c1, c2))
```

### Error message

```shell
c1: [0.02750148 0.0552023 ]
c2: [0.02753768 0.05533435]
type: float32 shape: (2,)
/Users/mep/anaconda3/envs/scipy-bug/lib/python3.12/site-packages/scipy/spatial/distance.py:1259: RuntimeWarning: invalid value encountered in sqrt
  return np.sqrt(js / 2.0)
first jsd: nan
second jsd: 0.0001306777430188037
```

### SciPy/NumPy/Python version and system information

```shell
The bug occurs on the new versions of scipy, numpy, python shown below, but it also happens for 1.10.1, 1.24.3, 3.8.15 respectively.


1.11.4 1.26.3 sys.version_info(major=3, minor=12, micro=1, releaselevel='final', serial=0)
/Users/mep/anaconda3/envs/scipy-bug/lib/python3.12/site-packages/scipy/__config__.py:146: UserWarning: Install `pyyaml` for better output
  warnings.warn("Install `pyyaml` for better output", stacklevel=1)
{
  "Compilers": {
    "c": {
      "name": "clang",
      "linker": "ld64",
      "version": "14.0.6",
      "commands": "x86_64-apple-darwin13.4.0-clang"
    },
    "cython": {
      "name": "cython",
      "linker": "cython",
      "version": "0.29.36",
      "commands": "cython"
    },
    "c++": {
      "name": "clang",
      "linker": "ld64",
      "version": "14.0.6",
      "commands": "x86_64-apple-darwin13.4.0-clang++"
    },
    "fortran": {
      "name": "gcc",
      "linker": "ld64",
      "version": "11.2.0",
      "commands": "/Users/builder/cbouss/perseverance-python-buildout/croot/scipy_1701810578094/_build_env/bin/x86_64-apple-darwin13.4.0-gfortran"
    },
    "pythran": {
      "version": "0.13.1",
      "include directory": "../../_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_p/lib/python3.12/site-packages/pythran"
    }
  },
  "Machine Information": {
    "host": {
      "cpu": "x86_64",
      "family": "x86_64",
      "endian": "little",
      "system": "darwin"
    },
    "build": {
      "cpu": "x86_64",
      "family": "x86_64",
      "endian": "little",
      "system": "darwin"
    },
    "cross-compiled": false
  },
  "Build Dependencies": {
    "blas": {
      "name": "mkl-sdl",
      "found": true,
      "version": "2023.1",
      "detection method": "pkgconfig",
      "include directory": "/Users/mep/anaconda3/envs/scipy-bug/include",
      "lib directory": "/Users/mep/anaconda3/envs/scipy-bug/lib",
      "openblas configuration": "unknown",
      "pc file directory": "/Users/mep/anaconda3/envs/scipy-bug/lib/pkgconfig"
    },
    "lapack": {
      "name": "mkl-sdl",
      "found": true,
      "version": "2023.1",
      "detection method": "pkgconfig",
      "include directory": "/Users/mep/anaconda3/envs/scipy-bug/include",
      "lib directory": "/Users/mep/anaconda3/envs/scipy-bug/lib",
      "openblas configuration": "unknown",
      "pc file directory": "/Users/mep/anaconda3/envs/scipy-bug/lib/pkgconfig"
    },
    "pybind11": {
      "name": "pybind11",
      "version": "2.10.4",
      "detection method": "pkgconfig",
      "include directory": "/Users/mep/anaconda3/envs/scipy-bug/include"
    }
  },
  "Python Information": {
    "path": "/Users/mep/anaconda3/envs/scipy-bug/bin/python",
    "version": "3.12"
  }
}
```

### PR DESCRIPTION
**Pull request by ghost**
Closes #20083

The jensenshannon function returns NaN values for some distributions close to each other by taking the square root of a negative number. The negative numbers appear in the return of the rel_entr function.
To fix this, use the equivalent kl_div function instead of rel_entr because it always returns non-negative values.
This is possible because the extra parameters on the kl_div formula that make the values non-negative will be canceled out in the sum.

### CODE REVIEW

File: scipy/spatial/tests/test_distance.py [hunk 1]
```diff
@@ -2246,6 +2246,9 @@ def test_jensenshannon():
                         [0.1954288, 0.1447697, 0.1138377, 0.0927636])
     assert_almost_equal(jensenshannon(a, b, axis=1),
                         [0.1402339, 0.0399106, 0.0201815])
+    assert_almost_equal(jensenshannon(np.array([0.027501475, 0.055202297], dtype='float32'),
```
Comment (j-bowhay) [inline scipy/spatial/tests/test_distance.py:2249]: new tests should use `assert_allclose`
```suggestion
    assert_allclose(jensenshannon(np.array([0.027501475, 0.055202297], dtype='float32'),
```

Comment (lucascolley) [reply scipy/spatial/tests/test_distance.py:2249]: maybe worth changing the rest of this test at the same time for consistency

Comment (lucascolley) [reply scipy/spatial/tests/test_distance.py]: Better to figure out the correct tolerances, maybe try `assert_allclose(..., rtol=1.5e-6)`. The default for `assert_allclose` is `rtol=1e-7`, while the default for `assert_array_almost_equal` is roughly `atol=1e-6`. The difference between `rtol` and `atol` can be quite significant.

**Comment by yuanx749** [comment]
Though I suggested this workaround in https://github.com/scipy/scipy/issues/20083#issuecomment-1980946457, I am not sure whether this is a reasonable solution.

In the example of the original issue #20083, the second jsd is:
```console
second jsd: 0.0001306777430188037
```
After this fix, the result becomes:
```console
second jsd: 0.0001788337319063894
```
Seems the results are unstable due to the small numbers. Thoughts from the maintainers would be appreciated.

**Pull request closed without merging**